lib/honeybadger/http_adapter.ex

```diff
@@ -0,0 +1,173 @@
+  @type headers :: [{binary(), binary()}]
+
+  @callback request(method(), binary(), body(), headers(), Keyword.t()) ::
+              {:ok, map()} | {:error, any()}
+
+  @doc """
+  Sets a user agent header.
+
+  The header value will be `Honeybadger-VERSION` with VERSION being the `:vsn` of
+  `:honeybadger` app.
+  """
+  @spec user_agent_header() :: {binary(), binary()}
+  def user_agent_header do
+    version = Application.spec(:honeybadger, :vsn) || "0.0.0"
+
+    {"User-Agent", "Honeybadger-#{version}"}
+  end
+
+  @default_http_client Enum.find_value(
+                         [
+                           {Req, Honeybadger.HTTPAdapter.Req},
+                           {Finch, Honeybadger.HTTPAdapter.Finch},
+                           {:hackney, Honeybadger.HTTPAdapter.Hackney}
+                         ],
+                         fn {dep, module} ->
+                           Code.ensure_loaded?(dep) && {module, []}
+                         end
+                       )
+
+  @doc """
+  Makes an HTTP request.
+
+  ## Options
+
+  - `:http_adapter` - The HTTP adapter to use, defaults to
+    `#{inspect(elem(@default_http_client, 0))}`
+  """
+  @spec request(atom(), binary(), binary() | nil, list(), Keyword.t()) ::
+          {:ok, HTTPResponse.t()} | {:error, HTTPResponse.t()} | {:error, term()}
+  def request(method, url, body, headers, opts) do
+    {http_adapter, http_adapter_opts} = get_adapter(opts)
+
+    method
+    |> http_adapter.request(url, body, headers, http_adapter_opts)
+    |> case do
+      {:ok, response} ->
+        decode_response(response, opts)
+
+      {:error, error} ->
+        {:error,
+         ServerUnreachableError.exception(
+           reason: error,
+           http_adapter: http_adapter,
+           request_url: url
+         )}
+    end
+    |> case do
+      {:ok, %{status: status} = resp} when status in 200..399 ->
+        {:ok, %{resp | http_adapter: http_adapter, request_url: url}}
+
+      {:ok, %{status: status} = resp} when status in 400..599 ->
+        {:error, %{resp | http_adapter: http_adapter, request_url: url}}
+
+      {:error, error} ->
+        {:error, error}
+    end
+  end
+
+  defp get_adapter(opts) do
+    default_http_adapter = Application.get_env(:honeybadger, :http_adapter, @default_http_client)
+
+    case Keyword.get(opts, :http_adapter, default_http_adapter) do
+      {http_adapter, opts} -> {http_adapter, opts}
+      http_adapter when is_atom(http_adapter) -> {http_adapter, nil}
+    end
+  end
+
+  @doc """
+  Decodes request response body.
+  """
+  @spec decode_response(HTTPResponse.t(), Keyword.t()) ::
+          {:ok, HTTPResponse.t()} | {:error, InvalidResponseError.t()}
+  def decode_response(%HTTPResponse{} = response, opts) do
+    case decode(response.headers, response.body, opts) do
+      {:ok, body} -> {:ok, %{response | body: body}}
+      {:error, _error} -> {:error, InvalidResponseError.exception(response: response)}
+    end
```

Comment:
Req already decodes responses automatically. This could be inlined into `hackney` if `finch` is removed.

Reply:
This makes a lot of sense, so I have moved the decoding code into the hackney adapter module. 👍